maven-embedder/src/main/java/org/apache/maven/cli/transfer/ConsoleMavenTransferListener.java

```diff
@@ -34,79 +36,159 @@
     }
 
     @Override
-    public synchronized void transferInitiated(TransferEvent event) {
-        overridePreviousTransfer(event);
-
-        super.transferInitiated(event);
+    public void transferInitiated(TransferEvent event) {
+        transfers.put(event.getResource(), event);
+        synchronized (lock) {
+            lock.notifyAll();
+        }
     }
 
     @Override
-    public synchronized void transferCorrupted(TransferEvent event) throws TransferCancelledException {
-        overridePreviousTransfer(event);
-
-        super.transferCorrupted(event);
+    public void transferCorrupted(TransferEvent event) throws TransferCancelledException {
+        transfers.put(event.getResource(), event);
+        synchronized (lock) {
+            lock.notifyAll();
+        }
     }
 
     @Override
-    public synchronized void transferProgressed(TransferEvent event) throws TransferCancelledException {
-        TransferResource resource = event.getResource();
-        transfers.put(resource, event.getTransferredBytes());
-
-        buffer.append("Progress (").append(transfers.size()).append("): ");
-
-        Iterator<Map.Entry<TransferResource, Long>> entries =
-                transfers.entrySet().iterator();
-        while (entries.hasNext()) {
-            Map.Entry<TransferResource, Long> entry = entries.next();
-            long total = entry.getKey().getContentLength();
-            Long complete = entry.getValue();
+    public void transferProgressed(TransferEvent event) throws TransferCancelledException {
+        transfers.put(event.getResource(), event);
+        synchronized (lock) {
+            lock.notifyAll();
+        }
+    }
 
-            String resourceName = entry.getKey().getResourceName();
+    @Override
+    public void transferSucceeded(TransferEvent event) {
+        transfers.put(event.getResource(), event);
+        synchronized (lock) {
+            if (transfers.size() == 1) {
+                doUpdate();
+            } else {
+                lock.notifyAll();
+            }
+        }
+    }
 
-            if (printResourceNames) {
-                int idx = resourceName.lastIndexOf('/');
+    @Override
+    public void transferFailed(TransferEvent event) {
+        transfers.put(event.getResource(), event);
+        synchronized (lock) {
+            if (transfers.size() == 1) {
+                doUpdate();
+            } else {
+                lock.notifyAll();
+            }
+        }
+    }
 
-                if (idx < 0) {
-                    buffer.append(resourceName);
-                } else {
-                    buffer.append(resourceName, idx + 1, resourceName.length());
+    void update() {
+        synchronized (lock) {
+            try {
+                long t0 = System.currentTimeMillis();
+                while (true) {
+                    lock.wait(100);
+                    long t1 = System.currentTimeMillis();
+                    if (t1 - t0 > MIN_DELAY_BETWEEN_UPDATES) {
```

Comment:
Don't use ? `currentTimeMillis()` for elapsed time. Use `System#nanoTime()`.

Reply:
I really don't need a resolution below 10ms, this is really not time sensitive at all.  I don't have any real problem changing, but what's the reasoning here ?

Reply:
I am not talking about resolution, but the fact that current millis is wall clock and can run backwards. `nanoTime()` is guaranteed to be at least monotonic.